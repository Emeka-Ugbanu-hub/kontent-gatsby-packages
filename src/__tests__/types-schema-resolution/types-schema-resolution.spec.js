
const { KontentTestHttpService }
  = require('@kentico/kontent-test-http-service-js');

const { createSchemaCustomization } = require('../../gatsby-node');

const fakeTypesResponse =
  require('./fakeTypesResponse.json');
const fakeItemsResponse =
  require('./fakeItemsResponse.json');
const fakeTaxonomiesReponse =
  require('../fakeTaxonomiesResponse.json');

describe(
  `Complex content types schema resolution`,
  () => {
    // Project ID 532c0844-5eb3-0033-7d1b-faeea1e7e407
    const fakeHttpServiceConfig = new Map();
    fakeHttpServiceConfig.set(
      /https:\/\/deliver.kontent.ai\/.*\/items/,
      {
        fakeResponseJson: fakeItemsResponse,
        throwError: false,
      });
    fakeHttpServiceConfig.set(
      /https:\/\/deliver.kontent.ai\/.*\/types/,
      {
        fakeResponseJson: fakeTypesResponse,
        throwError: false,
      });
    fakeHttpServiceConfig.set(
      /https:\/\/deliver.kontent.ai\/.*\/taxonomies/,
      {
        fakeResponseJson: fakeTaxonomiesReponse,
        throwError: false,
      });

    const createTypesMock = jest.fn();

    const mockedSchema = { buildObjectType: jest.fn((input) => (
      { data: `Returning from mocked schema's buildObjectType function.
        Original input: ${input}`,
      }))}; ;

    const api = {
      actions: {
        createTypes: createTypesMock,
      },
      schema: mockedSchema,
    };
    const deliveryClientConfig = {
      projectId: 'dummyProject',
      typeResolvers: [],
      httpService: new KontentTestHttpService(
        fakeHttpServiceConfig,
      ),
    };

    const pluginConfiguration = {
      deliveryClientConfig,
      languageCodenames: ['default'],
    };

    it('passes with no error', async () => {
      await createSchemaCustomization(api, pluginConfiguration);

      const createTypeCalls = createTypesMock.mock.calls;
      expect(createTypeCalls).toHaveLength(2);
      expect(createTypeCalls).toMatchSnapshot();

      const buildObjectTypeCalls = mockedSchema.buildObjectType.mock.calls;
      expect(buildObjectTypeCalls).toHaveLength(6);
      expect(buildObjectTypeCalls).toMatchSnapshot();
    });
  });
