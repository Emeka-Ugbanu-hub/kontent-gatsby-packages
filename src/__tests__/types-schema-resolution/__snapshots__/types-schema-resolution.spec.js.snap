// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Complex content types schema resolution passes with no error 1`] = `
Array [
  Array [
    "
    interface KontentType @nodeInterface {
      id: ID!
      system: KontentTypeSystem!
      elements: [KontentTypeElement]
      contentItems: [KontentItem] @link(by: \\"id\\", from: \\"contentItems___NODE\\")
    }
    type KontentTypeSystem @infer {
      codename: String!
      id: String!
      lastModified: Date! @dateformat
      name: String!
    }
    type KontentTypeElement @infer {
      codename: String!
      name: String!
      type: String!
      taxonomyGroup: String
      options: [ElementOption]
    }
    type ElementOption {
      name: String
      codename: String
    }
    interface KontentItem @nodeInterface {
      id: ID!
      system: KontentItemSystem!
      preferred_language: String!
      contentType: KontentType! @link(by: \\"id\\", from: \\"contentType___NODE\\")
      usedByContentItems: 
        [KontentItem] @link(by: \\"id\\", from: \\"usedByContentItems___NODE\\")
    }
    interface KontentElement @dontInfer {
      name: String!
      type: String!
    }
    type KontentItemSystem @infer {
      codename: String!
      id: String!
      language: String!
      lastModified: Date! @dateformat
      name: String!
      type: String!
    }
    type KontentAsset @infer {
      name: String!
      description: String
      type: String!
      size: Int!
      url: String!
      width: Int
      height: Int
    }
    type KontentElementMultipleChoiceValue @infer {
      codename: String!
      name: String!
    }
    type KontentElementRichTextResolvedData @infer{
      html: String
      linkedItemCodenames: [String]
      componentCodenames: [String]
    }
    type KontentRichTextImage @infer {
      description: String
      height: Int!
      imageId: String!
      url: String!
      width: Int!
    }
    type KontentRichTextLink @infer {
      codename: String!
      linkId: String!
      type: String!
      urlSlug: String!
    }
    type KontentAssetElement implements KontentElement @infer {
      name: String!
      type: String!
      value: [KontentAsset]
    }
    type KontentDateTimeElement implements KontentElement @infer {
      name: String!
      type: String!
      value: Date @dateformat
    }
    type KontentModularContentElement implements KontentElement @infer {
      name: String!
      type: String!
      itemCodenames: [String]
      linked_items: [KontentItem] @link(by: \\"id\\", from: \\"linked_items___NODE\\")
    }
    type KontentMultipleChoiceElement implements KontentElement @infer {
      name: String!
      type: String!
      value: [KontentElementMultipleChoiceValue]
    }
    type KontentNumberElement implements KontentElement @infer {
      name: String!
      type: String!
      value: Float
    }
    type KontentRichTextElement implements KontentElement @infer {
      name: String!
      type: String!
      value: String
      images: [KontentRichTextImage]
      links: [KontentRichTextLink]
      linked_items: [KontentItem] @link(by: \\"id\\", from: \\"linked_items___NODE\\")
      linkedItemCodenames: [String]
      resolvedData: KontentElementRichTextResolvedData
    }

    type KontentTaxonomyElement implements KontentElement @infer {
      name: String!
      type: String!
      taxonomyGroup: String!
      value: [KontentTaxonomyItem]
    }
    type KontentTaxonomyItem @infer {
      name: String!
      codename: String!
    }
    type KontentTextElement implements KontentElement @infer {
      name: String!
      type: String!
      value: String
    }
    type KontentUrlSlugElement implements KontentElement @infer {
      name: String!
      type: String!
      value: String
    }
    type KontentCustomElement implements KontentElement @infer {
      name: String!
      type: String!
      value: String
    }
  ",
  ],
  Array [
    Array [
      Object {
        "data": "Returning from mocked schema's buildObjectType function.
        Original input: [object Object]",
      },
      Object {
        "data": "Returning from mocked schema's buildObjectType function.
        Original input: [object Object]",
      },
      Object {
        "data": "Returning from mocked schema's buildObjectType function.
        Original input: [object Object]",
      },
      Object {
        "data": "Returning from mocked schema's buildObjectType function.
        Original input: [object Object]",
      },
      Object {
        "data": "Returning from mocked schema's buildObjectType function.
        Original input: [object Object]",
      },
      Object {
        "data": "Returning from mocked schema's buildObjectType function.
        Original input: [object Object]",
      },
    ],
  ],
]
`;

exports[`Complex content types schema resolution passes with no error 2`] = `
Array [
  Array [
    Object {
      "fields": Object {
        "contentItems": Object {
          "resolve": [Function],
          "type": "[KontentItem-withitem]",
        },
        "elements": "[KontentTypeElement]",
        "system": "KontentTypeSystem!",
      },
      "infer": false,
      "interfaces": Array [
        "Node",
        "KontentType",
      ],
      "name": "KontentType-withitem",
    },
  ],
  Array [
    Object {
      "fields": Object {
        "description": Object {
          "type": "KontentRichTextElement",
        },
        "title": Object {
          "type": "KontentTextElement",
        },
      },
      "infer": false,
      "name": "KontentItem-withitemElements",
    },
  ],
  Array [
    Object {
      "fields": Object {
        "contentType": Object {
          "resolve": [Function],
          "type": "KontentType-withitem!",
        },
        "elements": "KontentItem-withitemElements!",
        "preferred_language": "String!",
        "system": "KontentItemSystem!",
        "usedByContentItems": Object {
          "resolve": [Function],
          "type": "[KontentItem]",
        },
      },
      "infer": false,
      "interfaces": Array [
        "Node",
        "KontentItem",
      ],
      "name": "KontentItem-withitem",
    },
  ],
  Array [
    Object {
      "fields": Object {
        "contentItems": Object {
          "resolve": [Function],
          "type": "[KontentItem-withoutitem]",
        },
        "elements": "[KontentTypeElement]",
        "system": "KontentTypeSystem!",
      },
      "infer": false,
      "interfaces": Array [
        "Node",
        "KontentType",
      ],
      "name": "KontentType-withoutitem",
    },
  ],
  Array [
    Object {
      "fields": Object {
        "born": Object {
          "type": "KontentDateTimeElement",
        },
        "count_of_everything": Object {
          "type": "KontentNumberElement",
        },
        "custocm_color_picker": Object {
          "type": "KontentCustomElement",
        },
        "description": Object {
          "type": "KontentRichTextElement",
        },
        "photos": Object {
          "type": "KontentAssetElement",
        },
        "related_items": Object {
          "type": "KontentModularContentElement",
        },
        "title": Object {
          "type": "KontentTextElement",
        },
        "your_awesome_choices": Object {
          "type": "KontentMultipleChoiceElement",
        },
      },
      "infer": false,
      "name": "KontentItem-withoutitemElements",
    },
  ],
  Array [
    Object {
      "fields": Object {
        "contentType": Object {
          "resolve": [Function],
          "type": "KontentType-withoutitem!",
        },
        "elements": "KontentItem-withoutitemElements!",
        "preferred_language": "String!",
        "system": "KontentItemSystem!",
        "usedByContentItems": Object {
          "resolve": [Function],
          "type": "[KontentItem]",
        },
      },
      "infer": false,
      "interfaces": Array [
        "Node",
        "KontentItem",
      ],
      "name": "KontentItem-withoutitem",
    },
  ],
]
`;
